src/ratelimit/RateLimitController.test.ts

```diff
@@ -0,0 +1,132 @@
+import { ControllerMessenger } from '../ControllerMessenger';
+import {
+  ControllerActions,
+  RateLimitStateChange,
+  RateLimitController,
+  RateLimitMessenger,
+  GetRateLimitState,
+  CallApi,
+} from './RateLimitController';
+
+const name = 'RateLimitController';
+
+const implementations = { showNativeNotification: jest.fn() };
+
+type RateLimitedApis = typeof implementations;
+
+/**
+ * Constructs a unrestricted controller messenger.
+ *
+ * @returns A unrestricted controller messenger.
+ */
+function getUnrestrictedMessenger() {
+  return new ControllerMessenger<
+    GetRateLimitState<RateLimitedApis> | CallApi<RateLimitedApis>,
+    RateLimitStateChange<RateLimitedApis>
+  >();
+}
+
+/**
+ * Constructs a restricted controller messenger.
+ *
+ * @param controllerMessenger - An optional unrestricted messenger
+ * @returns A restricted controller messenger.
+ */
+function getRestrictedMessenger(
+  controllerMessenger = getUnrestrictedMessenger(),
+) {
+  return controllerMessenger.getRestricted<
+    typeof name,
+    ControllerActions<RateLimitedApis>['type'],
+    never
+  >({
+    name,
+    allowedActions: ['RateLimitController:call'],
+  }) as RateLimitMessenger<RateLimitedApis>;
+}
+
+const origin = 'snap_test';
+const message = 'foo';
+const successResult = { isRateLimited: false, result: undefined };
+
+describe('RateLimitController', () => {
+  jest.useFakeTimers();
```

Comment:
This will enable fake timers for the entire test suite, globally, as soon as this line is run. Is that what we want? Isn't this enabled by default anyway?

Reply:
I think this is in line with Jest's recommendations for their [timer mocks](https://jestjs.io/docs/timer-mocks), and [fake timers are not enabled by default](https://jestjs.io/docs/configuration#timers-string).

Reply:
That's intended and I don't believe it's enabled by default 👍

Reply:
This doesn't seem great to me. Now the behavior of other test suite depends on the order in which Jest runs them. Sounds like a nightmare to debug if that ever impacts them.

The Jest docs suggest calling `useRealTimers` afterwards if you want to avoid impacting other tests. That seems safer to me, less likely to impact other tests. They can be enabled in `beforeEach` and disabled in `afterEach`. That also ensures the fake timers get reset in between each test, ensuring tests don't unintentionally end up depending on state from the previous test.

Reply:
Sure!